data/tutorials/language/0lg_09_env.md

```diff
@@ -0,0 +1,562 @@
+
+The classic (and very inefficient) way to present recursion is using the function that computes numbers of the [Fibonacci Sequence](https://en.wikipedia.org/wiki/Fibonacci_sequence).
+```ocaml
+# let rec fibo n = if n <= 1 then n else fibo (n - 1) + fibo (n - 2);;
+val fibo : int -> int = <fun>
+
+# List.init 10 Fun.id |> List.map fibo;;
+- : int list = [0; 1; 1; 2; 3; 5; 8; 13; 21; 34]
+```
+
+This version of `fibo` is inefficient because the number of recursive calls created doubles at each call, which creates exponential growth.
+
+This version does a better job:
+```ocaml
+# let rec fibo m n i = if i = 0 then m else fibo n (n + m) (i - 1);;
+val fibo : int -> int -> int -> int = <fun>
+
+# let fibo = fibo 0 1;;
+val fibo : int -> int = <fun>
+
+# List.init 10 Fun.id |> List.map fibo;;
+- : int list = [0; 1; 1; 2; 3; 5; 8; 13; 21; 34]
+```
+
+The first version takes two extra parameters: the two previously computed Fibonacci numbers.
+
+The second version uses the two first Fibonacci numbers as initial values. There is nothing to be computed when returning from a recursive call, so this enables the compiler to perform an optimisation called “tail call elimination,” which turns recursivity into imperative iteration in the generated native code and leads to much improved performances.
+
+#### Passing Multiple Arguments to Functions
+
+Until now, functions with several arguments had a type looking like this: `a₁ -> a₂ -> a₃ -> ⋯ -> b`
+
+Where `a₁` is the type of the first argument, `a₂` is the type of the second argument, etc., and `b` is the type of the result. However, this type isn't displayed as it is. In reality, this type is the following: `a₁ -> (a₂ -> (a₃ -> ⋯ -> b))`. The arrow symbol is a binary operator. Here is how this can be observed:
+```ocaml
+# ( + );;
+- : int -> int -> int = <fun>
+
+# let f : (int -> (int -> int)) = ( + );;
+val f : int -> int -> int = <fun>
+```
+
+In this example, the type of the integer addition function `( + )` is displayed first. It prints as `int -> int -> int`. It takes two integers and returns an integer. In the second part, `f` is defined to have type `int -> (int -> int)`. Despite the type of `( + )` looking different, the binding with `f` is allowed because those two types are the same. The toplevel always prints the version without parentheses to make it easier to read.
+
+Putting the parentheses the other way does not work:
+```ocaml
+# let g : ((int -> int) -> int) = ( + );;
+Error: This expression has type int -> int -> int
+       but an expression was expected of type (int -> int) -> int
+       Type int is not compatible with type int -> int
+```
+
+In mathematical language, it is said that the type arrow operator _associates to the right_. Function types without parentheses should be thought to have parentheses put to the right like the type of `f` was declared. These types are the same:
```

Comment:
Would it be correct to say: 
"...put to the right, like when the type of `f` was declared" ?
or 
"...put to the right, like how the type of `f` was declared"

Is this referring to the `let f : (int -> (int -> int)) = ( + );;` above? 
Do you mean put the parentheses to the right like the `(int -> ....) in this example?


What do you think of this construction for the first part?
"Function types without parentheses should be treated as if they have parentheses, placed to the right..."

Reply:
Yes it is possible and indeed better

Reply:
Suggestion committed

Reply:
I'll update it, and then. you can see if it reads better and still accurate! 🙂